### ISSUE TITLE
Python API for directories

### ISSUE THREAD

**Issue by spirali** [substantic/rain#10]
# Directories - Python API

This is a proposal for directory manipulation API. Directory can be considered as a tarball blob with content type "dir". The only special "magic" happens when such data object is mapped to a file system,
then it is unpacked as a directory. It also works in the opposite direction, when a data object of content type "dir" is collected from the file system, it has to be directory (in contarst to a file in case of non-dir content type). Of course usual optizations for moving dataobject inside a worker is applied, hence
there is no obsolete packing/unpacking to from/tar tar when it is not needed.

## Constructor of constant data 

The wrapper over blob (as for example 'pickled') that creates a constant data object:

```python
directory(path="/abc")  # Creates a data object from client's local directory
directory(dict={"fileX": b"xxxx", "fileY": b"yyyy" })  # Creates a directory dataobject from "dict" description.
```

## Getting subparts

Now let "d" be a directory dataobject, then following tasks creates a new dataobject from a file/directory inside "d".

```python
tasks.get_path(d, "subdir/file1.txt", content_type="json")
tasks.get_path(d, "subdir", content_type="dir")
```

I propose to define a new term "fragment" that is a pointer inside a directory. The following code creates a fragment:

```python
x = d.get_pah("subdir/file1.txt", content_type="json")
```

Fragment can be used as dataobject, e.g.:

```python
t = tasks.concat((x, x))
```

But creating a fragment does not generate a task, the task consuming fragment directly gets a subpart of the original object. (It utilizes to name a subdir in inputs of a task).

## Write method

The part of the proposed API is also "write" method on DataIntance that writes an object to a local file system. If data object is normal blob, it creates a file, if it is a directory it is extracted as a directory to filesystem.

```python
d.fetch().write("/path")
```

## Make directory

Task that creates directory from given data objects:

```python
tasks.make_directory({"subdir": {"file1": a }})
```

**Comment by spirali** [comment]
With small changes implemented in #14 (without fragments)

**Comment by spirali** [comment]
Directory API now seems complete, except two aspects: data types in DO that are tracked at #18 and fragments. After some face-to-face discussions, fragments looks quite redundant and problematic API (or at least it needs more time to thinking). Therefore, I am closing this generic thread. Feel free to open new issue for any specific part of directory API.
